sklearn/feature_selection/mutual_info_.py

```diff
@@ -0,0 +1,447 @@
+from scipy.sparse import issparse
+from scipy.special import digamma
+
+from ..externals.six import moves
+from ..metrics.cluster.supervised import mutual_info_score
+from ..neighbors import NearestNeighbors
+from ..preprocessing import scale
+from ..utils import check_random_state
+from ..utils.validation import check_X_y
+from ..utils.multiclass import check_classification_targets
+
+
+def _compute_mi_cc(x, y, n_neighbors):
+    """Compute mutual information between two continuous variables.
+
+    Parameters
+    ----------
+    x, y : ndarray
+        Samples of random variables, 1-d arrays of identical shape.
+
+    n_neighbors : int
+        Number of nearest neighbors to search for each point, see [1]_.
+
+    Returns
+    -------
+    mi : float
+        Estimated mutual information. If it turned out to be negative it is
+        replace by 0.
+
+    Notes
+    -----
+    True mutual information can't be negative. If its estimate by a numerical
+    method is negative, it means (providing the method is adequate) that the
+    mutual information is close to 0 and replacing it by 0 is a reasonable
+    strategy.
+
+    References
+    ----------
+    .. [1] A. Kraskov, H. Stogbauer and P. Grassberger, "Estimating mutual
+           information". Phys. Rev. E 69, 2004.
+    """
+    n_samples = x.size
+
+    x = x.reshape((-1, 1))
+    y = y.reshape((-1, 1))
+    xy = np.hstack((x, y))
+
+    nn = NearestNeighbors(metric='chebyshev', n_neighbors=n_neighbors)
+
+    nn.fit(xy)
+    radius = nn.kneighbors()[0]
+    radius = np.nextafter(radius[:, -1], 0)
+
+    nn.set_params(algorithm='kd_tree')
+
+    nn.fit(x)
+    ind = nn.radius_neighbors(radius=radius, return_distance=False)
+    nx = np.array([i.size for i in ind])
+
+    nn.fit(y)
+    ind = nn.radius_neighbors(radius=radius, return_distance=False)
+    ny = np.array([i.size for i in ind])
+
+    mi = (digamma(n_samples) + digamma(n_neighbors) -
+          np.mean(digamma(nx + 1)) - np.mean(digamma(ny + 1)))
+
+    return max(0, mi)
+
+
+def _compute_mi_cd(c, d, n_neighbors):
+    """Compute mutual information between continuous and discrete variables.
+
+    Parameters
+    ----------
+    c : ndarray
+        Samples from a continuous random variable.
+
+    d : ndarray
+        Samples from a discrete random variable.
+
+    n_neighbors : int
+        Number of nearest neighbors to search for each point, see [1]_.
+
+    Returns
+    -------
+    mi : float
+        Estimated mutual information. If it turned out to be negative it is
+        replace by 0.
+
+    Notes
+    -----
+    True mutual information can't be negative. If its estimate by a numerical
+    method is negative, it means (providing the method is adequate) that the
+    mutual information is close to 0 and replacing it by 0 is a reasonable
+    strategy.
+
+    References
+    ----------
+    .. [1] B. C. Ross "Mutual Information between Discrete and Continuous
+       Data Sets". PLoS ONE 9(2), 2014.
```

Comment:
I decided not to include this reference, because it seems impossible to find this paper in English anywhere. Nevertheless, the method is well known and it is explained well in Kraskov et al.

Basically PLoS paper explains how to apply this established method to our problem. I think it's a relevant and useful reference.